include/seqan3/range/decorator/gap_decorator.hpp

```diff
@@ -1038,6 +656,389 @@ template <std::ranges::view urng_t>
 gap_decorator(urng_t range) -> gap_decorator<urng_t>;
 //!\}
 
+/*!\brief A special iterator that enforces std::random_access_iterator albeit it's complexity is logarithmic.
+ *
+ * \details
+ *
+ * \attention This iterator models std::random_access_iterator but **does not guarantee constant time access**.
+ *            Random access is provided in \f$O(\log k)\f$ time (where `k` is the number of gaps).
+ *
+ * This iterator returns values when dereferenced, not references, i.e. it does not model the
+ * [Cpp17random_access_iterator](https://en.cppreference.com/w/cpp/named_req/random_access_iterator)
+ * requirements of the STL. It does model the C++20 std::random_access_iterator.
+ *
+ * You can get a view over the gap_decorator with this random access iterator by using
+ * seqan3::views::pseudo_random_access. But note, that after adapting the gap_decorator the non-range interfaces
+ * are lost (e.g. insert_gap or erase_gap). This adaptor allows to use the gap_decorator as a range in algorithms
+ * like std::ranges::distance and get the more efficient logarithmic implementation. The regular iterator only
+ * supports std::bidirectional_iterator and thus linear time would be needed to calculated the distance.
+ *
+ * \see seqan3::gap_decorator::gap_decorator_iterator
+ */
+template <std::ranges::viewable_range inner_type>
+//!\cond
+    requires std::ranges::random_access_range<inner_type> && std::ranges::sized_range<inner_type> &&
+             (std::is_const_v<std::remove_reference_t<inner_type>> || std::ranges::view<inner_type>)
+//!\endcond
+class gap_decorator<inner_type>::gap_decorator_iterator_ra
```

Comment:
Any reason to move this outside the class or just preference?

Reply:
in general the class itself becomes really hard to read with large nested classes inside of it. So we started to outsource the implementation in other places as well.

Reply:
If we do this now we should make this official since others need to know. 
I added a discussion card. We should be consistent with our code style.